src/serialize/numpy.rs

```diff
@@ -603,3 +636,115 @@ impl<'p> Serialize for NumpyBool {
         serializer.serialize_bool(self.value)
     }
 }
+
+/// This mimicks the
+#[derive(Clone, Copy)]
+pub enum NumpyDatetimeUnit {
+    Years,
+    Months,
+    Weeks,
+    Days,
+    Hours,
+    Minutes,
+    Seconds,
+    Milliseconds,
+    Microseconds,
+    Nanoseconds,
+    Picoseconds,
+    Femtoseconds,
+    Attoseconds,
+    Generic,
+}
+
+const ONE_SECOND_IN_NANOS: i64 = 1_000_000_000;
+
+impl NumpyDatetimeUnit {
+    /// Create a `NumpyDatetimeUnit` from a pointer to a Python object holding a
+    /// numpy array.
+    ///
+    /// This function must only be called with pointers to numpy arrays.
+    ///
+    /// We need to look inside the `obj.dtype.descr` attribute of the Python
+    /// object rather than using the `descr` field of the `__array_struct__`
+    /// because that field isn't populated for datetime64 arrays; see
+    /// https://github.com/numpy/numpy/issues/5350.
+    fn from_pyobject(ptr: *mut PyObject) -> Self {
+        let dtype = ffi!(PyObject_GetAttr(ptr, DTYPE_STR));
+        let descr = ffi!(PyObject_GetAttr(dtype, DESCR_STR));
+        let el0 = ffi!(PyList_GET_ITEM(descr, 0));
+        let descr_str = ffi!(PyTuple_GET_ITEM(el0, 1));
+        let mut str_size: pyo3::ffi::Py_ssize_t = 0;
+        let uni = crate::unicode::read_utf8_from_str(descr_str, &mut str_size);
+        let fmt = str_from_slice!(uni, str_size);
+        match fmt[4..].trim_end_matches(']') {
+            "Y" => Self::Years,
+            "M" => Self::Months,
+            "W" => Self::Weeks,
+            "D" => Self::Days,
+            "h" => Self::Hours,
+            "m" => Self::Minutes,
+            "s" => Self::Seconds,
+            "ms" => Self::Milliseconds,
+            "us" => Self::Microseconds,
+            "ns" => Self::Nanoseconds,
+            "ps" => Self::Picoseconds,
```

Comment:
Also if picoseconds is not generally used, fine, it's OK to bail out like with non-C contiguous arrays. The complexity in numpy is enough for compromises in support like this.

Reply:
Cool, that's implemented in the latest commit.